Add `allowed_cluster_replica_sizes` system variable

aalexandrov opened the pull request:
This adds the `allowed_cluster_replica_sizes` system variable proposed in #15208

### Motivation

  * This PR adds a known-desirable feature.

   Closes #15208.

### Tips for reviewer

CC-ing @benesch, but I think @jkosh44 has sufficient knowledge in these parts to do the review, so I only need one approval from both of you for the adapter changes. Otherwise nothing special in terms of tips. I tried to split commits in semantic units, so I suggest doing the review one commit at a time.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes (thew new system variable is not modifiable by end users at the moment).

jkosh44 commented on src/adapter/src/session/vars.rs:
```diff
@@ -1428,6 +1449,31 @@ impl Value for [String] {
+impl Value for Vec<String> {
+    const TYPE_NAME: &'static str = "string array";
+
+    fn parse(s: &str) -> Result<Vec<String>, ()> {
```
I may have misunderstood, but I thought we decided against this syntax: https://github.com/MaterializeInc/materialize/issues/15208#issuecomment-1283318785?

aalexandrov replied:
I opened the PR with the state from yesterday evening and only saw the updated message after that. I will have to extend the parser before we can merge this.

benesch commented:
Mentioned this on the issue but I think the comment got lost: I have a local branch from a while back with the parser changes + half of the implementation to support multiple entries in `search_path`. I can dig it up tonight.

aalexandrov commented:
> Mentioned this on the issue but I think the comment got lost: I have a local branch from a while back with the parser changes + half of the implementation to support multiple entries in `search_path`. I can dig it up tonight.

I added both `Literals()` and `Idents()` to `SetVariableValueItem` and changed the parser accordingly. The former is used for `allowed_cluster_replica_sizes`, while the latter can be used for `search_path`. @benesch if you point me to your WIP I'll make sure to wrap up the path thread in a separate PR.

aalexandrov commented on src/sql-parser/src/parser.rs:
```diff
@@ -4540,13 +4556,47 @@ impl<'a> Parser<'a> {
-        Ok(match (self.parse_value(), token) {
-            (Ok(value), _) => SetVariableValue::Literal(value),
-            (Err(_), Some(Token::Keyword(DEFAULT))) => SetVariableValue::Default,
-            (Err(_), Some(Token::Keyword(kw))) => SetVariableValue::Ident(kw.into_ident()),
```
@jkosh44 / @benesch any idea why this is needed? For example
```sql
set search_path to select;
```
is accepted by Materialize, but is rejected by Postgres, so it's not an issue fo Postgres-compatibility.

aalexandrov replied:
OK, I think this might have something to do with it:
```sql
SET standard_conforming_strings = ON;
```

jkosh44 replied:
Postgres and us allow the value to not have quotes around it in certain cases. For example we both allow

```SQL
SET TRANSACTION_ISOLATION TO serializable;
```
Without that line of code we'd reject all values without quotes. Clearly though we're a bit too lenient and allow too many unquoted values compared to Postgres. I can open an issue for this. Is it a blocker for this PR?

aalexandrov replied:
> SET standard_conforming_strings = typo;

This will be covered by the `Token::Ident` case. I think the problem is that some unquoted strings (such as `ON`) will be tokenized as a `Token::Keyword` instead of a `Token::Identifier`, which is why you need the `SetVariableValue::Ident(kw.into_ident())` case.

aalexandrov commented on src/adapter/src/session/vars.rs:
```diff
@@ -1428,6 +1451,41 @@ impl Value for [String] {
+    const TYPE_NAME: &'static str = "string list";
+
+    fn parse(s: &str) -> Result<Vec<String>, ()> {
+        match parse_set_variable_value(s) {
```
I think we should save ourselves a non-trivial amount of duplicated effort for implementing parsing and formatting logic for various `Value` implementations if we just delegate that to the `Display` and `Parser` logic of the `SetVariableValue` structure that models the corresponding `Value` type.

jkosh44 replied:
That seems reasonable. The coupling exist whether or not we make it explicit, in that we'll have to always keep the two implementations in sync. We might as well make it explicit.

aalexandrov commented:
FYI, it seems that some part of the logic conflates

1. the presentation of `string`-like `Value` instances as strings, which when rendered in `SHOW variable_name` should be rendered without the surrounding quotes, with 
2. the conversion of `Value` instances as strings in order to store them in the stash (and subsequently parse them with `Value::parse`).

It seems to me that the code will be a bit simpler if we extended `Value` as follows:

```rust
/// A value that can be stored in a session variable.
pub trait Value: ToOwned + Send + Sync {
    /// The name of the value type.
    const TYPE_NAME: &'static str;
    /// Parses a value of this type from a string.
    fn parse(s: &str) -> Result<Self::Owned, ()>;
    /// Formats this value as a string for storage purposes.
    fn format(&self) -> String;
    /// Formats this value as a string for display purposes.
    fn display(&self) -> String {
        self.format()
    };
}
```

aalexandrov commented:
> Can you give some examples? I'm having trouble fully understanding.

TL;DR: this is really outside the scope of this PR, so feel free to ignore completely.

It seems that `Value::parse` might be used only for (1). I can only see `Var::value` being called from for `ServerVar` / `SystemVar` / `SessionVar`, and as far as I can see `Var::value` is called only for presentation purposes (`sequence_show_all_variables`, `sequence_show_variable`, `AdapterError::fmt`).

I was expecting at least one of the `Var::value` calls to result in persisted information somewhere in the catalog. However, at least at first glance it seems that this should be the result of handling `catalog::Op::UpdateSystemConfiguration` instance somewhere down the line in `Coordinator::catalog_transact`, and those instances are created exclusively in `sequence_alter_system_set` without relying on `Var::value`.

It will help me if I can understand who and where persists the catalog commands. From what I can see this is done in the `tx.upsert_system_config(&name, &value)?;` for the `Op::UpdateSystemConfiguration` case in `Catalog::transact_inner`.

If all of the above is correct, maybe we can get rid of the quote-stripping here:

```rust
let op = match value {
    SetVariableValue::Default => catalog::Op::ResetSystemConfiguration { name },
    SetVariableValue::Literal(Value::String(value)) => {
        catalog::Op::UpdateSystemConfiguration { name, value }
    }
    SetVariableValue::Ident(ident) => catalog::Op::UpdateSystemConfiguration {
        name,
        value: ident.into_string(),
    },
    value => catalog::Op::UpdateSystemConfiguration {
        name,
        value: value.to_string(),
    },
};
```

by removing the second and third case and adjusting the `Value::parse` function accordingly.

aalexandrov commented:
I will open a separate PR to the cloud repository after testing with `kind` tomorrow.

Note that I named the CLI parameter (and the associated config paramters) `bootstrap-system-vars` as opposed to `bootstrap-set`, as the latter is a bit vague, especially given the fact that we have other CLI & config paramters with a `bootstrap` prefix.

jkosh44 commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -373,6 +373,14 @@ pub struct Args {
         default_value = "1"
     )]
     bootstrap_builtin_cluster_replica_size: String,
+    /// An optional colon-separated list of $var_name=$var_value pairs for
```
I think it's semi-colon-separated, right?

jkosh44 reviewed (commented):
Mostly LGTM, but we need some tests with the new system variable and the command line args.

aalexandrov commented:
I added tests in `test/cloud`:

```bash
./mzcompose --dev run test-resource-limits
./mzcompose --dev run test-bootstrap-vars
```

jkosh44 reviewed (approved):
LGTM

benesch commented:
Some post merge feedback here:

  * Though we use the term "variable" internally, PostgreSQL is consistent in calling these "configuration parameters." So anything that appears in the user interface of Materialize (e.g., command-line flags) should use the term "configuration parameter" or "parameter" rather than "variable."
  * The library interface of `environmentd` (i.e., `mz_environmentd::Config`) should take fully-parsed values. Any string manipulation should happen in the binary crate (src/environmentd/src/bin/environmentd/main.rs). Specifically, the type of `bootstrap_system_vars` should be `HashMap<String, String>`, or somesuch, not `Option<String>`.
  * We have prior art for command-line parameters that take multiple KEY=VALUE pairs. See, e.g., `--orchestrator-kubernetes-service-label`. Any new command-line arguments that take multiple KEY=VALUE pairs should be consistent with these existing command-line arguments.

benesch commented:
I put up https://github.com/MaterializeInc/materialize/pull/16046 to address the above.
